test/test_serialization.py

```diff
@@ -1007,6 +1007,501 @@ def _save_load_check(t):
         with self.assertRaisesRegex(RuntimeError, err_msg):
             _save_load_check(t)
 
+    def test_serialization_byteorder_mark(self):
+        lstm = torch.nn.LSTM(3, 3)
+        inputs = [torch.randn(1, 3) for _ in range(5)]
+        inputs = torch.cat(inputs).view(len(inputs), 1, -1)
+        hidden = (torch.randn(1, 1, 3), torch.randn(1, 1, 3))  # clean out hidden state
+
+        databuffer = io.BytesIO()
+        torch.save(lstm.state_dict(), databuffer)
+        databuffer.seek(0)
+
+        with torch.serialization._open_zipfile_reader(databuffer) as zip_file:
+            byteordername = 'byteorder'
+            self.assertTrue(zip_file.has_record(byteordername))
+            byteorderdata = zip_file.get_record(byteordername)
+            self.assertTrue(byteorderdata in [ b'little', b'big' ])
+            self.assertEqual(byteorderdata.decode(), sys.byteorder)
+
+    def test_serialization_invalid_byteorder_mark(self):
+        lstm = torch.nn.LSTM(3, 3)
+        inputs = [torch.randn(1, 3) for _ in range(5)]
+        inputs = torch.cat(inputs).view(len(inputs), 1, -1)
+        hidden = (torch.randn(1, 1, 3), torch.randn(1, 1, 3))  # clean out hidden state
+
+        # set invalid byteorder value
+        current_byteorder = sys.byteorder
+        sys.byteorder = 'invalid'
+
+        databuffer = io.BytesIO()
+        torch.save(lstm.state_dict(), databuffer)
```

Comment:
We should probably throw the error in `torch.save` as well. If we're going to forbid `torch.load`ing a file that has a byteorder that is not either 'little' or 'big', then I don't think we should allow generating such files in the first place

Reply:
This is related to https://github.com/pytorch/pytorch/pull/94503#discussion_r1160143799